mandible/metadata_mapper/directive/directive.py

```diff
@@ -0,0 +1,51 @@
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from typing import ClassVar, Dict, Optional, Type
+
+from ..context import Context
+from ..source import Source
+from ..types import Key
+
+DIRECTIVE_REGISTRY: Dict[str, Type["TemplateDirective"]] = {}
+
+
+def get_key(key: Key, context: Context) -> str:
+    if callable(key):
+        key = key(context)
+
+    return key
+
+
+@dataclass
+class TemplateDirective(ABC):
+    """Base class for directives in a metadata template.
+
+    A directive is a special marker in the metadata template which will be
+    replaced by the MetadataMapper.
+    """
+    # Registry boilerplate
+    def __init_subclass__(
+        cls,
+        register: bool = True,
+        name: Optional[str] = None,
+        **kwargs,
+    ):
+        if register:
+            name = name or cls.__name__.lower()
+            DIRECTIVE_REGISTRY[name] = cls
+            cls.directive_name = name
+
+        super().__init_subclass__(**kwargs)
+
+    # Begin class definition
+    directive_name: ClassVar[Optional[str]] = None
```

Comment:
Should this be below source since it has a default vaule?

Reply:
No. It's a class var so it won't be part of the generated constructor. It is set automatically in the `__init_subclass__` as part of the registry mechanism.

Reply:
Pulling the code into my IDE, I am getting a syntax error stating, `Fields with a default value must come after any fields without a default.`

Reply:
Looking into why that is the case.

Reply:
Perhaps some pycharm lint that doesn't understand how `ClassVar` works with dataclasses. Flake8 doesn't have any issues with it and it clearly works because there are unit tests that import this class and even check the `directive_name` attribute.

Reply:
Ya, must be. After a quick search, I couldn't see anything saying it was wrong. All the examples I did see did seem to still put the default value at the end though.

Reply:
I do think that putting a parameter with a default first in this case, is inconsistent with how we do it everywhere else, and unless there is a technical reason behind why, we should be consistent with other scenarios

Reply:
I don't want to mix class vars with regular attributes. It will only cause confusion.

Reply:
What is the harm in replacing
```python
    # Begin class definition
    directive_name: ClassVar[Optional[str]] = None

    context: Context
    sources: Dict[str, Source]
```
with
```python
    # Begin class definition
    context: Context
    sources: Dict[str, Source]
    directive_name: ClassVar[Optional[str]] = None
```
?
This seems entirely inconsequential other than appeasing Pycharm.

Reply:
I feel like it makes it harder to see that one of the variables is a class var. Doesn't it make the most sense to put class level stuff at the top and then instance level stuff further down? That's also why I like to put the `__init_subclass__` at the top.

Reply:
Yeah. Actually because it's a class variable Pycharm shouldn't be having a fit about it. It seems like there's a convention of putting class level stuff before instance level stuff that I'm not aware of. I wasn't tracking this well at first. Seems like maybe we should create a new coding convention for the standards repo. Maybe we can come to some middle ground to fix this issue with Pycharm, for instance we can disable that inspection that's flagging with:
```python
    # noinspection PyDataclass
    directive_name: ClassVar[Optional[str]] = None
```
but this is also IDE-specific, but maybe that's okay because it's not unreasonable to imagine a future where other devs work on this using Pycharm and run into the same issue we're having.

Reply:
I can see your logic. I don't think it would make it any harder since it says `ClassVar` on that line of code. At the end of the day, our standards do not state anything about this case, and while seeing the red line will be annoying, put it where it feels right for you.